[Nguyen Minh Hieu] iP

hieunm1821 opened the pull request:
# DukePro

> " Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's 
- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2122s2.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :wink:

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are Java programmer, you can use it to practice Java too. Here's the main method: 
```java
    public static void main(String[] args) {
        new Duke().run();
    }
```

Commit:
Add Gradle support

hieunm1821 committed:
Add Greet, echo, exit

hieunm1821 committed:
Adding function, listing function

hieunm1821 committed:
Add Task class, mark as Done function

hieunm1821 committed:
Add toString in Task, fix index when mark/unmark

hieunm1821 committed:
ToDos, Events, Deadlines

hieunm1821 committed:
Fix bug when print Todo

hieunm1821 committed:
Add semi-automate testing

hieunm1821 committed:
Rewrite main into small function

hieunm1821 committed:
Handle errors

hieunm1821 committed:
Add delete function and use ArrayList to store Tasks

hieunm1821 committed:
Change savedTask to plural form

hieunm1821 committed:
Delete indentation for case clauses in switch

hieunm1821 committed:
Add class Storage and Parser

hieunm1821 committed:
Use LocalDate for Deadline

hieunm1821 committed:
Merge branch 'branch-Level-7'

hieunm1821 committed:
Merge branch 'branch-Level-8'

hieunm1821 committed:
Add class ExceptionHandler

hieunm1821 committed:
Fix Deadline datetime error

hieunm1821 committed:
More OOP

hieunm1821 committed:
Fix printing Mark/Unmark

hieunm1821 committed:
Divide classes into packages

hieunm1821 committed:
Add JUnit tests

hieunm1821 committed:
Create a JAR file

hieunm1821 committed:
Add JavaDoc comments

hieunm1821 committed:
Tweak the code to comply with a coding standard

hieunm1821 committed:
Add finding function

hieunm1821 committed:
Changes in data file

hieunm1821 committed:
Merge branch 'branch-A-JavaDoc'

hieunm1821 committed:
Merge branch 'branch-A-CodingStandard'

hieunm1821 committed:
Merge branch 'branch-Level-9'

hieunm1821 committed:
Add comments for new function

hieunm1821 committed:
Changes to non-static methods

Geralddtan commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,63 @@
+        while (true) {
+            String userInput = ui.readCommand(sc);
+            try {
+                Command command = parser.parse(userInput);
+                command.execute(tasks, ui, storage);
```
I like how clean and concise the parsing and execution of commands are!

Geralddtan commented on src/main/java/duke/component/Ui.java:
```diff
@@ -0,0 +1,111 @@
+
+import duke.task.Task;
+
+import static duke.constant.Message.GOOD_BYE;
+import static duke.constant.Message.GREETING;
+import static duke.constant.Message.HORIZONTAL_LINE;
+import static duke.constant.Message.LINE_PREFIX;
+import static duke.constant.Message.LINE_SEPARATOR;
+import static duke.constant.Message.CONFIRM_MARK;
+import static duke.constant.Message.CONFIRM_UNMARK;
```
I really like your use of constants as it helps to organise your code much better!

Geralddtan commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,41 @@
+ * A class represents for an event.
+ */
+public class Event extends Task{
+    protected String at;
```
Perhaps a slightly more intuitive variable name here?
The variable "at" is not as descriptive as it could be! Perhaps you could use a variable like "details" instead if it makes sense!

Geralddtan commented on src/main/java/duke/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,60 @@
+            throw new MissingDateTimeException();
+        }
+
+            Task currentTask = new Deadline(deadlineDetail[0], deadlineDetail[1]);
+            tasks.addTask(currentTask);
+
+            ui.printConfirmAdd(currentTask, tasks.getNumberOfTasks());
+            storage.writeTaskToFile(tasks);
```
Perhaps this block of code should be one indentation (4 spacings) to the left to match the indentation of the "if" clause!

Geralddtan reviewed (commented):
Overall, I found your code very easy to read and very organised, this made it very easy to follow your code. The use of constants is especially great. You also followed the coding standard really closely, making it hard for me to find many coding standard violations. LGTM, just a few nits to fix.

hieunm1821 committed:
Merge branch 'add-gradle-support'

hieunm1821 committed:
Use Gradle

jaegarpoon commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,63 @@
+     * Class constructor
+     * Initialize tasks by reading data file
+     */
+    public Duke() {
```
I like how most functions are abstracted out!

jaegarpoon commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,23 @@
+     * @throws DukeException if Duke-specified exception
+     * @throws Exception if other exception
+     */
+    public abstract void execute(TaskList tasks, Ui ui, Storage storage) throws DukeException, Exception;
```
I like how the execute function manages the input for each command

jaegarpoon commented on src/main/java/duke/constant/CommandConstant.java:
```diff
@@ -0,0 +1,16 @@
+/**
+ * A class that contains constants related to Command.
+ */
+public final class CommandConstant {
```
I like how you use constants for all related fields, but perhaps you can consider enums?

jaegarpoon reviewed (commented):
Overall good work! Enjoyed reviewing your code with just some minor comments.

hieunm1821 committed:
Add basic GUI

hieunm1821 committed:
Merge branch 'branch-Level-10'

hieunm1821 committed:
Improve code quality

Code have some code quality issues.

Improve code quality helps to improve understandability

Let's fix code quality issues

hieunm1821 committed:
Add bubble around text, add greeting

hieunm1821 committed:
Add assertions

Code lacks assertions

Assertions help to verify the program state in the runtime.

Let's add some assertion in some command class.

hieunm1821 committed:
Merge pull request #1 from hieunm1821/branch-A-Assertions

Add assertions

hieunm1821 committed:
Merge branch 'master' into branch-A-Assertions

hieunm1821 committed:
Merge branch 'master' into branch-A-CodeQuality

hieunm1821 committed:
Merge pull request #2 from hieunm1821/branch-A-CodeQuality

Improve code quality

hieunm1821 committed:
Merge branch 'master' into branch-A-CodeQuality

hieunm1821 committed:
Set up CI

hieunm1821 committed:
Change location of gradle.yml

hieunm1821 committed:
Add LocalDateTime for Event, Deadline

Cannot treat the input time as just a String

Let's use LocalDateTime to parse input.

hieunm1821 committed:
Merge branch 'branch-A-CodeQuality'

hieunm1821 committed:
Merge branch 'branch-A-Assertions'

hieunm1821 committed:
Add viewing schedule on specific date

hieunm1821 committed:
Merge pull request #3 from hieunm1821/branch-B-ViewSchedules

Branch b view schedules

hieunm1821 committed:
Tweak the GUI

hieunm1821 committed:
Set title for window and add screenshot of product

hieunm1821 committed:
Update User Guide

hieunm1821 committed:
Edit User Guide

hieunm1821 committed:
Release jar
